storage-transcribe-audio/functions/src/index.ts

```diff
@@ -1,9 +1,141 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
 import * as functions from "firebase-functions";
+import * as admin from "firebase-admin";
+import { getEventarc, Channel } from "firebase-admin/eventarc";
+import * as speech from "@google-cloud/speech";
+import * as path from "path";
+import * as os from "os";
+
+import * as logs from "./logs";
+import {
+  publishFailureEvent,
+  errorFromAny,
+  publishCompleteEvent,
+} from "./util";
+import mkdirp = require("mkdirp");
+import {
+  transcodeToLinear16AndUpload,
+  transcribeAndUpload,
+} from "./transcribe-audio";
+
+admin.initializeApp();
+
+const eventChannel: Channel | null = process.env.EXT_SELECTED_EVENTS
+  ? getEventarc().channel(process.env.EXT_SELECTED_EVENTS, {
+      allowedEventTypes: process.env.EXT_SELECTED_EVENTS,
+    })
+  : null;
+
+logs.init();
+
+const client = new speech.SpeechClient();
+
+// TODO(reao): Write to firestore if that setting is enabled
+// TODO(reao): Don't write to storage if that setting is not enabled
+export const transcribeAudio = functions.storage
+  .object()
+  .onFinalize(async (object): Promise<void> => {
+    logs.start();
+    const { contentType } = object; // the MIME type
+
+    if (object.metadata && object.metadata.isTranscodeOutput === "true") {
+      logs.audioAlreadyProcessed();
+      return;
+    }
+
+    if (!contentType) {
+      logs.noContentType();
+      return;
+    }
+
+    if (!contentType.startsWith("audio/")) {
+      logs.contentTypeInvalid(contentType);
+      return;
+    }
+
+    if (object.name === undefined) {
+      logs.undefinedObjectName(object);
+      return;
+    }
+
+    const bucket = admin.storage().bucket(object.bucket);
+    const filePath = object.name;
+
+    try {
+      const localCopyPath: string = path.join(os.tmpdir(), filePath);
+      const tempLocalDir = path.dirname(localCopyPath);
+
+      logs.tempDirectoryCreating(tempLocalDir);
+      await mkdirp(tempLocalDir);
+      logs.tempDirectoryCreated(tempLocalDir);
+
+      const remoteFile = bucket.file(filePath);
+      logs.audioDownloading(filePath);
+      await remoteFile.download({ destination: localCopyPath });
+      logs.audioDownloaded(filePath, localCopyPath);
+
+      const transcodeResult = await transcodeToLinear16AndUpload(
+        {
+          localCopyPath,
+          storageOutputPath: "out.wav",
+        },
+        bucket
+      );
+
+      if (transcodeResult.state === "failure") {
+        logs.transcodingFailed(transcodeResult);
+        eventChannel &&
+          (await publishFailureEvent(eventChannel, transcodeResult));
+        return;
+      }
+
+      const { sampleRateHertz, audioChannelCount } = transcodeResult;
+      const {
+        uploadResponse: [file /* metadata */],
+      } = transcodeResult;
+
+      const transcriptionResult = await transcribeAndUpload({
```

Comment:
The warnings from the transcoder are forgotten here; I have a commit ready to fix that so that they are kept.

Reply:
Done.